misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt

```diff
@@ -0,0 +1,382 @@
+                  FROM mysql.general_log
+                  WHERE command_type = 'Query'
+                  AND argument LIKE '%/* _stream%'
+                  ORDER BY event_time DESC
+                  LIMIT 1 OFFSET 1
+                """.trimIndent())
+            .uniqueResult { it.getString(1) }
+      }
+    }
+  }
+
+  /**
+   * Connects directly to the Docker Vitess mysqld, bypassing vtgate entirely. We use this to dig
+   * into the query log. This is a perpetual, not connection pooled connection so should not be
+   * closed. We shut down the Vitess docker container after the tests have completed running so
+   * this doesn't need to be closed explicitly.
+   */
+  private fun connect(): Connection {
+    var connection = connection;
+    if (connection != null) return connection;
+
+    val cluster = startVitessService.cluster()
+    connection = cluster.openMysqlConnection()
+    this.connection = connection
+    return connection
+  }
+
+  /**
+   * Figure out how many total full scatter queries we've executed so far.
+   */
+  private fun extractScatterQueryCount(): Int {
+    val request = Request.Builder()
+        .url("http://localhost:27000/debug/vars")
+        .build()
+    val response = okHttpClient.newCall(request).execute()
+    val source = response.body()!!.source()
+    val adapter = moshi.adapter<Variables>()
+    val variables = adapter.fromJson(source)!!
+    return variables.QueriesProcessed["SelectScatter"] ?: 0
+  }
+
+  /**
+   * We don't use this anymore but we may want to introspect the query plans again in the future
+   * so I'm leaving it around for now.
+   */
+  private fun extractScatterQueryCountFromPlan(): Int {
+    val request = Request.Builder()
+        .url("http://localhost:27000/debug/query_plans")
+        .build()
+    val response = okHttpClient.newCall(request).execute()
+    val source = response.body()!!.source()
+    return parseQueryPlans(source).filter { it.isScatter }.sumBy { it.ExecCount }
+  }
+
+  private val EMPTY_LINE = "\n\n".encodeUtf8()
+
+  internal fun parseQueryPlans(data: BufferedSource): Sequence<QueryPlan> {
+    // Read (and discard) the "Length" line
+    data.readUtf8Line()
+
+    val adapter = moshi.adapter<QueryPlan>()
+
+    return data.split(EMPTY_LINE).map { buffer ->
+      // Discard top line
+      buffer.readUtf8Line()
+      try {
+        adapter.fromJson(buffer)
+      } catch (e: EOFException) {
+        null
+      }
+    }.filterNotNull()
+  }
+
+  fun <T> disable(body: () -> T): T = disabled.withValue(true, body)
+}
+
+private inline fun <T, R> ThreadLocal<T>.withValue(value: T, body: () -> R): R {
+  val prev = get()
+  set(value)
+  try {
+    return body()
+  } finally {
+    set(prev)
+  }
+}
+
+/**
+ * Returns a list containing elements after the last element that satisfied the given [predicate]
+ * or the empty list if no elements satisfy the predicate.
+ */
+inline fun <T> List<T>.takeLastFrom(predicate: (T) -> Boolean): List<T> {
```

Comment:
when I read this function name I assume it’s going to mutate the List, but it doesn’t. Maybe sublistAfterMatching ?

Reply:
Yeah I didn't know what to name it either 👍